File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        # Lets redo this question with binary search 
        fullArray = []
        for rows in range(len(matrix)):
            for cols in range(len(matrix[0])):
                fullArray.append(matrix[rows][cols])
        
        
        #binary search 
        lo = 0
        hi = len(fullArray)-1
        
        
        while lo<=hi:
            pivot = (lo+hi)//2
            if fullArray[pivot]<target:
                lo+=1
            elif fullArray[pivot]>target:
                hi-=1
            else:
                return True
        return False
            

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        //1. Determine rows and cols count
        int noRows = matrix.length;
        int noCols = matrix[0].length;
        
        //2. Do binary search 
        int left = 0;
        int right = noRows * noCols -1;
        
        while (left<=right){
            int pivotIndex = (int)((left + right)/2);
            int pivotElement = matrix[(int)(pivotIndex/noCols)][pivotIndex%noCols];
            
            if(target == pivotElement){
                return true;
            }else{
                if (target<pivotElement){
                    right = pivotIndex-1;
                }else{
                    left = pivotIndex +1;
                }
            }
        }
        
        return false;
    }
}